File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
// 1번 노드에서 가장 멀리 떨어진 노드의 갯수
// 다익스트라
// 최대값 몇개인지
// n 2 ~ 20000
// m 1 ~ 50000
// 다익스트라 45만

import java.util.*;

class Solution {
    public int solution(int n, int[][] edge) {
        int answer = 0;
        
        ArrayList<Integer>[] graph = new ArrayList[n+1];
        for (int i = 0; i <= n ; i++) {
            graph[i] = new ArrayList<Integer>();
        }
        for (int[] e : edge) {
            int a = e[0], b = e[1];
            graph[a].add(b);
            graph[b].add(a);
        }
        
        int[] res = new int[n+1];
        Arrays.fill(res, Integer.MAX_VALUE);
        res[1] = 0;
        PriorityQueue<int[]> pq = new PriorityQueue<>((o1, o2) -> o1[0] - o2[0]); 
        pq.add(new int[]{0, 1});
        while (!pq.isEmpty()) {
            int[] q = pq.poll();
            int dis = q[0], node = q[1];
            
            if (dis < res[node]) continue;
            
            for (int i = 0; i < graph[node].size(); i++){
                if (res[graph[node].get(i)] > dis + 1) {
                    res[graph[node].get(i)] = dis+1;
                    pq.add(new int[]{dis+1, graph[node].get(i)});
                }
            }
            
        }
        
        // 최대값 구하기
        int max = 0;
        for (int i = 1; i < n+1 ; i++ ){
            if (max < res[i])
                max = res[i];
        }
        
        // 최대값 개수 세기
        for (int i = 1; i < n+1 ; i++ ){
            if (res[i] == max){
                answer++;
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
# 다익스트라
# 50000 * log 20000

import heapq

def solution(n, edge):
    answer = 0
    
    graph = [[] for _ in range(n+1)]
    d = [float("inf")] * (n+1)
    for a, b in edge:
        graph[a].append(b)
        graph[b].append(a)
    
    q = []
    d[1] = 0
    heapq.heappush(q, (0, 1))
    
    while q:
        distance, node = heapq.heappop(q)
        if distance > d[node]:
            continue
            
        for next in graph[node]:
            if d[next] > distance + 1:
                d[next] = distance + 1
                heapq.heappush(q, (distance + 1, next))
    
    maxD = max(d[1:])
    for r in d:
        if r == maxD:
            answer+=1
    
    return answer